[READY] Add an option to make relative paths in flags from extra conf absolute

micbou opened the pull request:
Since PR https://github.com/Valloric/ycmd/pull/680, all the infrastructure is here to convert any relative paths in the list of flags from the `.ycm_extra_conf.py` file to absolute ones. We just need to use the `_MakeRelativePathsInFlagsAbsolute` function.

This PR adds a new optional key `working_directory` to the dictionary returned by the `FlagsForFile` function in the `.ycm_extra_conf.py` file. This option allows specifying the working directory i.e. the directory for which the paths in the list of flags are relative to.

Update YCM's own `.ycm_extra_conf.py` file to reflect this change.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/valloric/ycmd/795)
<!-- Reviewable:end -->

micbou commented:
Looks like Travis recently added the `bin` folder in the home directory. Linux builds are now failing with this error:
```
mkdir: cannot create directory ‘/home/travis/bin’: File exists
```
We need to merge PR #793 first to add the `-p` flag to [the `mkdir` command in our Travis script](https://github.com/Valloric/ycmd/blob/master/ci/travis/travis_install.linux.sh#L6). I can also create a separate PR for that.

---

Reviewed 4 of 4 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpHIdPaoZR0Dk3aBeU1:b-yv52dz)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I worry that this is possibly a breaking change by default. By manipulating the paths when the new flag is not specified we are introducing silent change in behaviour, which I worry will break existing (working) installations. I am fairly certain that this would break my setup at work for example, where the relative nature of the include paths is accommodated in the make system integration. 

---

Review status: 3 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpI4i5ZbE6hrEdHnfWu:b-vskvt5)*
<!-- Sent from Reviewable.io -->

micbou commented:
This only affects relative paths. Are you passing relative paths to libclang? What's the expected behavior in that case? Seems undefined to me.

---

Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpI5eZqn-JmQed1YsCw:bcilya7)*
<!-- Sent from Reviewable.io -->

puremourning commented:
I'm pretty sure we pass -I. and it is interpreted relative to the TU. Memory could be off in that. Passing relative paths to a compiler is pretty normal behaviour. I suppose I t might be interpreted relative to the wd of ycmd, but (especially in the case of a global extra conf), I can't see a good justification for assuming that the YCM extra conf path as the root of such relative paths.

 The relative paths in a standard cc invocation are interpreted relative to the wd of the compiler when invoked which is (let's say) typically the path of the cc file. In the compilation database case, we set the calls relative to the compiler wd entry in the database, not relative to the compile commands file.  

Ofc every project is different, but that's why I'm nervous about making assumptions about users' project structure or paradigms. 

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpI731R6bDjjiNvmg5z:b1vydb8)*
<!-- Sent from Reviewable.io -->

Valloric commented:
I think @puremourning is right; this would break some people. We could perhaps do this without the new default of always resolving relative paths; we'd only do it when `working_directory` is provided.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpIHjzxD5J7EdXZeGPN:b1ssgjf)*
<!-- Sent from Reviewable.io -->

micbou commented:
> I'm pretty sure we pass -I. and it is interpreted relative to the TU.

Seems really convoluted to write include paths in its `.ycm_extra_conf.py` relative to the TU, i.e. the currently edited file.

> I can't see a good justification for assuming that the YCM extra conf path as the root of such relative paths.

Because that's natural. When writing relative paths in a file, users expect them to be relative to this file.

> I think @puremourning is right; this would break some people. We could perhaps do this without the new default of always resolving relative paths; we'd only do it when working_directory is provided.

This defeats the purpose of this PR which is helping users writing `.ycm_extra_conf.py` files, because they won't know about this option. They will write relative paths in their `.ycm_extra_conf.py` files, of course relative to the `.ycm_extra_conf.py` file, and it won't work.

I'd also like to point out that, in addition to being less than obvious (what is the actual working directory?), the current behavior for relative paths is undocumented. Do we want to go as far as not breaking backward compatibility for undocumented and obscure behavior?

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpIA1MTxsd4f-JipHop:b2qqq8e)*
<!-- Sent from Reviewable.io -->

puremourning commented:
The use case in question is where the flags are not actually writen into the `.ycm_extra_conf.py`, but the python within shells out to the make system to get the flags (in our case, effectively `make COMPILER=clang show_CXXFLAGS`). This returns the compile flags that would be passed for `file_name` when using the normal make system, rather than somehow repeating that logic within the extra conf file. Pretty sure we're not the only people which do this sort of thing, and it's essentially how the `compile_commands.json` works.

In any case, the current behaviour actually relative to the ycmd wd:

Given the following file structure:

```
$> find .
.
./test.h
./path
./path/test.h
./path/test.cc
./.ycm_extra_conf.py
ukfd-ltp2l(benj:MMA_main.git) tmp/relative>
```

with:

./test.h = `int test;`
./path/test.h = `int test;`
./path/test.cc = 

```c++
#include <test.h>
#include <stdio.h>

int main()
{
    printf( "%c", test );
}
```

and `.ycm_extra_conf.py`:

```python
def FlagsForFile( file_name, **kwargs ):
    return { 
        'flags': [
            '-x', 'c++',
            '-I.'
        ]
    }
```

* With Vim wd = ./path: `:YcmCompleter GetType` on `test` returns `char` (not `int`).
* With Vim wd = ./: `YcmCompleter GetType` on `test` returns `int`.

I confess that it (currently) being relative to ycmd working directory is arbitrary. However, I'm still not convinced that making it relative to the `.ycm_extra_conf.py` seems any less wrong. Again, in particular, when using a global `ycm_extra_conf.py` (which would be in any random dir).

If we were to pick a "default", I would suggest making it relative to the `dirname` of the `file_name`. Though this again is somewhat of a breaking change.

To be clear: I'm _totally_ up for adding the `working_dir` item to the dict. This is a good move and I will likely implement it here as `os.path.dirname( file_name )`, but I'm still worried that forcefully changing the paths to an (arbitrarily) different default, is just causing potentially avoidable headaches in the (perhaps unlikely) case that people have become accustomed to this behaviour. I think we can say in our docs that the default is ycmd working dir, as that's what it currently is.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpI9NNeqXvM7hd4jvDh:b-q6hwxh)*
<!-- Sent from Reviewable.io -->

micbou commented:
You convinced me. I'll update the PR to only make the paths absolute when `working_directory` is specified (is `working_dir` a better name?). That's a little unfortunate for new users. I think they would expect paths to be relative to a local `.ycm_extra_conf.py` by default (as you said, this doesn't make sense for the global extra conf). But we can't break users that rely on the current behavior.

---

Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpLECa_FRhAUSraFIK0:bk7rgut)*
<!-- Sent from Reviewable.io -->

puremourning commented:
Maybe we should straight up call a spoon a spoon and call it `paths_relative_to` as we aren't actually wetting a working dir?

Valloric commented:
Not at all against a clearer name than `working_dir`; we could call it `include_paths_relative_to_dir` and remove all confusion.

> This defeats the purpose of this PR which is helping users writing .ycm_extra_conf.py files, because they won't know about this option. 

I'd say the proportion of users who learn how to write an extra conf file by taking ycmd's (which we link to from the docs) and then change it to suit them is >95%. :D 

So if we use it, they'll know about the option.

---

Review status: 0 of 4 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KpNFkJLNU_5-ERs41JT:b-chiyna)*
<!-- Sent from Reviewable.io -->

micbou commented:
Renamed the option to `include_paths_relative_to_dir` and updated the `.ycm_extra_conf.py` file in the `examples` folder too. PR is ready.

---

Reviewed 4 of 4 files at r3, 4 of 4 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KqIELoPOYKM_r1CPbfx:b-52eyxz)*
<!-- Sent from Reviewable.io -->

vheon commented:
Reviewed 2 of 4 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795)*
<!-- Sent from Reviewable.io -->

Valloric commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks!

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KqsZImVycEnhbbA1y3S:b9benhj)*
<!-- Sent from Reviewable.io -->

bstaletic commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>
Seems to work as expected. Thanks.
Seeing that @puremourning was the last one asking questions, I'll let him trigger @zzbot.

---

Reviewed 2 of 4 files at r4, 1 of 1 files at r5, 2 of 2 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-Kr-0Nh8j7OcRljIikD6:bqc3fp8)*
<!-- Sent from Reviewable.io -->

puremourning commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Thanks!

@zzbot r+

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/valloric/ycmd/795#-:-KsdDgrSsQuLz0XXQfmv:b-7vhnyt)*
<!-- Sent from Reviewable.io -->
